### PR TITLE
feat(httpserver):  add new router

### PR DESCRIPTION
**Pull request by aniaan**
# benchmark

## Preparing

(4core/32GB/) * 4

| Name        | port  | vm   | version              |
| ----------- | ----- | ---- | -------------------- |
| ordered-v1  | 10080 | vm1   | main(golang 1.18.7) |
| ordered-v2  | 10080 | vm2   | feat/router(golang 1.18.7) |
| radixtree   | 10080 | vm3   | feat/router(golang 1.18.7)         |
| hey         | -     | vm4   | v0.1.4              |

[case](https://github.com/aniaan/eg-router-benchmark.git)

Start up the easegress server on vm1, vm2, vm3 and create the `mock-pipeline`

```shell
egctl object create -f eg-router-benchmark/static/mock-pipe.yaml
```

## Summary

The performance of `ordered-v1` and `ordered-v2` is basically the same, and the performance of `radixtree` is much better than them.

static: `radixtree` is about 3 times the performance of `ordered`

prefix: `radixtree` is about 3 times the performance of `ordered`

regexp: `radixtree` is about 40 times the performance of `ordered`

## static

The test case for static routes will create 10000 routes with md5 values from 1 to 10000, and test requests will be accessed using md5 values of 5000.

### 1. ordered-v1

vm1:

```shell
egctl object create -f eg-router-benchmark/static/order.yaml
```

**stress test**

vm4:

url:  `http://vm-1:10080/a35fe7f7fe8217b4369a0af4244d1fca`

eg: `./hey -n 900 -c 50 -m GET http://vm-1:10080/a35fe7f7fe8217b4369a0af4244d1fca`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 24068.3219   |
| 90000  | 100 | 26077.3742   |
| 90000   | 120 | 27219.1946   |

### 2. ordered-v2

vm2:

```shell
egctl object create -f eg-router-benchmark/static/order.yaml
```

**stress test**

vm4:

url:  `http://vm-2:10080/a35fe7f7fe8217b4369a0af4244d1fca`

eg: `./hey -n 900 -c 50 -m GET http://vm-2:10080/a35fe7f7fe8217b4369a0af4244d1fca`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 25078.4517   |
| 90000  | 100 | 27391.6867   |
| 90000   | 120 | 27401.6701   |

### 3. radixtree

vm3:

```shell
egctl object create -f eg-router-benchmark/static/redixtree.yaml
```

**stress test**

vm4:

url:  `http://vm-3:10080/a35fe7f7fe8217b4369a0af4244d1fca`

eg: `./hey -n 900 -c 50 -m GET http://vm-3:10080/a35fe7f7fe8217b4369a0af4244d1fca`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 41687.5741   |
| 90000  | 100 | 65122.1171   |
| 90000   | 120 | 65638.4786   |

## prefix

The test case for prefix-matching routes will create 10000 routes with md5 values from 1 to 10000, and test requests will be accessed using md5 values of 5000

### 1. ordered-v1

vm1:

```shell
egctl object create -f eg-router-benchmark/prefix/order.yaml
```

**stress test**

vm4:

url:  `http://vm-1:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

eg: `./hey -n 900 -c 50 -m GET http://vm-1:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 22110.7111   |
| 90000  | 100 | 26023.5309   |
| 90000   | 120 | 26016.2430   |

### 2. ordered-v2

vm2:

```shell
egctl object create -f eg-router-benchmark/prefix/order.yaml
```

**stress test**

vm4:

url:  `http://vm-2:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

eg: `./hey -n 900 -c 50 -m GET http://vm-2:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 25449.3521  |
| 90000  | 100 | 28119.3687   |
| 90000   | 120 | 27882.9445   |

### 3. radixtree

vm3:

```shell
egctl object create -f eg-router-benchmark/prefix/redixtree.yaml
```

**stress test**

vm4:

url:  `http://vm-3:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

eg: `./hey -n 900 -c 50 -m GET http://vm-3:10080/a35fe7f7fe8217b4369a0af4244d1fca/suffix`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 36541.2653   |
| 90000  | 100 | 58698.0033   |
| 90000   | 120 | 65060.4261   |

## regexp

The test case for regular matching routes will create 10000 RESTful style routes with the template `/users/{username}/<i>` and the test requests will be accessed using the md5 value of 5000

### 1. ordered-v1

vm1:

```shell
egctl object create -f eg-router-benchmark/regexp/order.yaml
```

**stress test**

vm4:

url:  `http://vm-1:10080/users/aniaan/5000`

eg: `./hey -n 900 -c 50 -m GET http://vm-1:10080/users/aniaan/5000`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 1769.1021   |
| 90000  | 100 | 1743.7057   |
| 90000   | 120 | 1743.2940   |

### 2. ordered-v2

vm2:

```shell
egctl object create -f eg-router-benchmark/regexp/order.yaml
```

**stress test**

vm4:

url:  `http://vm-2:10080/users/aniaan/5000`

eg: `./hey -n 900 -c 50 -m GET http://vm-2:10080/users/aniaan/5000`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 1733.3634   |
| 90000  | 100 | 1707.0785   |
| 90000   | 120 | 1686.6228   |

### 2. radixtree

vm3:

```shell
egctl object create -f eg-router-benchmark/regexp/radixtree.yaml
```

**stress test**

vm4:

url:  `http://vm-3:10080/users/aniaan/5000`

eg: `./hey -n 900 -c 50 -m GET http://vm-3:10080/users/aniaan/5000`

| requests    | concurrency  | tps   |
| ----------- | ----- | ---- |
| 900         | 50    | 38619.2734   |
| 90000  | 100 | 66155.8978   |
| 90000   | 120 | 66357.7316   |

### CODE REVIEW

File: pkg/object/httpserver/routers/order/router.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	path = mp.pathRE.ReplaceAllString(path, mp.RewriteTarget)
+	r.SetPath(path)
+}
+
+func (r *OrderRouter) Search(context *routers.RouteContext) {
+	req := context.Request
+	ip := req.RealIP()
+	path := context.Path
+
+	for _, host := range r.rules {
```
Comment (localvar) [inline pkg/object/httpserver/routers/order/router.go:137]: it is confusing to name the variable `host`.

File: pkg/object/httpserver/routers/spec.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+	if !p.matchable {
+		return true
+	}
+
+	// method match
+	req := context.Request
+	ip := req.RealIP()
+
+	if !p.AllowIP(ip) {
```
Comment (localvar) [inline pkg/object/httpserver/routers/spec.go:195]: IP should be checked after method, header, and query.

File: pkg/object/httpserver/routers/spec.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	Backend           string         `json:"backend" jsonschema:"required"`
+	ClientMaxBodySize int64          `json:"clientMaxBodySize" jsonschema:"omitempty"`
+	Headers           Headers        `json:"headers" jsonschema:"omitempty"`
+	Queries           Queries        `json:"queries,omitempty" jsonschema:"omitempty"`
+	MatchAllHeader    bool           `json:"matchAllHeader" jsonschema:"omitempty"`
+	MatchAllQuery     bool           `json:"matchAllQuery" jsonschema:"omitempty"`
+
+	ipFilter             *ipfilter.IPFilter
+	method               MethodType
+	cacheable, matchable bool
```
Comment (localvar) [inline pkg/object/httpserver/routers/spec.go:68]: propose not to use these two fields, they bring little performance benefit but make the code hard to read and error prone.

File: doc/reference/routers.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
+# Routers
+
+- [Routers](#routers)
+  - [Order](#order)
+  - [Art](#art)
+
+Router determines how requests are routed to the corresponding Pipeline for subsequent processing. We currently support two routing policies, `Order` and `Art`, and you can choose a Router that suits your needs based on its characteristics, and we also provide the ability to customize Router, if the built-in Router does not meet your needs, you can choose Write a custom Router.
```
Comment (localvar) [inline doc/reference/routers.md:7]: `policy` and `strategy` are both used in this doc for the same purpose. I think `strategy` is better, but please double confirm.

**Review by localvar — APPROVED** [review]

**Review by suchen-sci — COMMENTED** [review]
Is rules like `"/123/123/123/123{name}"` allowed? Is not allowed, you may need add some validation for the parameter match...

**Comment by suchen-sci** [comment]
If a request path match several rules, should we give the result by the order in definition?

For example
```
...
routerKind: RadixTree
rules:
  - paths:
    - path: /user/{username}/123
      backend: h1
    - path: /user/wang/{num}
      backend: h2
    - path: /{type}/wang/123
      backend: h3
...
```
for path `/user/wang/123` it match all three paths, should we return the first one define in yaml file, which should be `/user/{username}/123`? currently it returns `/user/wang/{num}`...

**Comment by aniaan** [comment]
> Is rules like `"/123/123/123/123{name}"` allowed? Is not allowed, you may need add some validation for the parameter match...

This is allowed

**Comment by aniaan** [comment]
> /user/wang/{num}

This is because the matching order of the prefix tree is decided, `full path -> regexp path -> parameter path -> prefix path`, so it will return `/user/wang/{num}`, RadixTree, compared with Ordered, is lost its definition order, so the user needs to understand the rules of RadixTree before using RadixTree.

**Comment by suchen-sci** [comment]
ok, know i fully understand this, is this case, match order is `/user/wang/{num}` > `/user/{username}/123` > `/{type}/wang/123` for `/user/wang/123`. 

So if a path matches several rules, we can split a path into chars, for every char follow the order of `full path -> regexp path -> parameter path -> prefix path`.

```
...
routerKind: RadixTree
rules:
  - paths:
    - path: /users/{u:[a-z]+}sername/hovercard
      backend: h1
    - path: /users/u{sername}/hovercard
      backend: h2
...
```
so in this case, `/users/username/hovercard` will match `/users/u{sername}/hovercard`. 

Very amazing work, so could you please update the doc, maybe add more examples to show this char level matching priority? there maybe some users make the same mistake like me.

**Comment by aniaan** [comment]
> ok, know i fully understand this, is this case, match order is `/user/wang/{num}` > `/user/{username}/123` > `/{type}/wang/123` for `/user/wang/123`.
> 
> So if a path matches several rules, we can split a path into chars, for every char follow the order of `full path -> regexp path -> parameter path -> prefix path`.
> 
> ```
> ...
> routerKind: RadixTree
> rules:
>   - paths:
>     - path: /users/{u:[a-z]+}sername/hovercard
>       backend: h1
>     - path: /users/u{sername}/hovercard
>       backend: h2
> ...
> ```
> 
> so in this case, `/users/username/hovercard` will match `/users/u{sername}/hovercard`.
> 
> Very amazing work, so could you please update the doc, maybe add more examples to show this char level matching priority? there maybe some users make the same mistake like me.

The document has been updated, you can re-read it

File: doc/reference/routers.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+      backend: h1
+    - path: /user/wang/{num}
+      backend: h2
+    - path: /{type}/wang/123
+      backend: h3
+```
+
+`/user/wang/123` will match `/user/wang/{num}`, which is determined by the matching order of RadixTree, you can split the path into individual characters, each character follows the matching order of `Full match` -> `Regexp match` -> `Parameter match` -> `Prefix match`, so the final match result is `/user/wang/{num}`.
+
+RadixTree, compared with Ordered, is lost its definition order.
```
Comment (xxx7xxxx) [inline doc/reference/routers.md:161]:
```suggestion
RadixTree, compared with Ordered, has lost its definition order.
```

File: doc/reference/routers.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+| path | Match backend | Rewrite target|
+|------|--------------|----------------|
+|/users/aniaan/hovercard | `full-match-backend` | /api/users/aniaan/card |
+|/users/12345/hovercard | `parameter-path-backend` | /api/users/12345/card |
+|/users/localvar/hovercard | `regexp-backend` |  /api/users/localvar/card |
+|/users/test | `prefix-backend` | /api/users/test |
+
+You can see that only the wildcard `*` needs to be replaced by the built-in variable `eg_wildcard`.
```
Comment (xxx7xxxx) [inline doc/reference/routers.md:215]: To be distinguished for built-in variables, how about using all capital letters as `EG_WILDCARD`

File: pkg/object/httpserver/routers/radixtree/router.go [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+		routers.Path
+		pattern         string
+		paramKeys       []string
+		rewriteTemplate *template.Template
+	}
+
+	paths []*muxPath
+
+	// Represents node and edge in radix tree
+	node struct {
```
Comment (xxx7xxxx) [inline pkg/object/httpserver/routers/radixtree/router.go:50]: Why not use and extend mature third vendors such as https://github.com/julienschmidt/httprouter. Are our requirements hard to satisfy by second-developing existing vendors?

Comment (aniaan) [reply pkg/object/httpserver/routers/radixtree/router.go:50]: The code used is actually from https://github.com/go-chi/chi/blob/master/tree.go and the documentation explains

File: pkg/object/httpserver/routers/radixtree/router.go [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+	"strings"
+	"text/template"
+
+	"github.com/megaease/easegress/pkg/object/httpserver/routers"
+	"github.com/megaease/easegress/pkg/util/stringtool"
+)
+
+// art-router implementation below is a based on the original work by
+// go-chi in https://github.com/go-chi/chi/blob/master/tree.go
+// (MIT licensed). It's been heavily modified for use as a HTTP router.
```
Comment (xxx7xxxx) [inline pkg/object/httpserver/routers/radixtree/router.go:34]:
```suggestion
// (MIT licensed). It's been heavily modified for use as an HTTP router.
```

File: pkg/object/httpserver/routers/radixtree/router.go [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+	"regexp"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/megaease/easegress/pkg/object/httpserver/routers"
+	"github.com/megaease/easegress/pkg/util/stringtool"
+)
+
+// art-router implementation below is a based on the original work by
```
Comment (xxx7xxxx) [inline pkg/object/httpserver/routers/radixtree/router.go:32]: What is an `art-router`?

Comment (aniaan) [reply pkg/object/httpserver/routers/radixtree/router.go]: History named, has been corrected

File: doc/reference/routers.md (hunk 6 above)
Comment (aniaan) [reply doc/reference/routers.md]: updated

**Review by xxx7xxxx — APPROVED** [review]

**Pull request merged**